docs/custom_promise_types/groups.py

```diff
@@ -0,0 +1,340 @@
+        
+        return repairs, failed_repairs
+
+    
+    def _exclude_users(self, group, users):
+        repairs = 0
+        failed_repairs = 0
+
+        for user in users:
+            self.log_debug(f"user '{user}' should be excluded from group '{group.name}'")
+            if user in group.members:
+                self.log_debug(f"user '{user}' not excluded from group '{group.name}'")
+                try:
+                    group.remove_member(user)
+                except GroupException as e:
+                    self.log_error(f"failed to remove user '{user}' from group '{group.name}': {e}")
+                    failed_repairs += 1
+                else:
+                    self.log_info(f"removing user '{user}' from group '{group.name}'")
+                    repairs += 1
+            else:
+                self.log_debug(f"user '{user}' already excluded from group '{group.name}'")
+
+        return repairs, failed_repairs
+
+    
+    def _only_users(self, group, users):
+        repairs = 0
+        failed_repairs = 0
+
+        self.log_debug(f"group '{group.name}' should only contain members {users}")
+        if (set(users) != set(group.members)):
+            self.log_debug(f"group '{group.name}' does not only contain members {users}")
+            try:
+                group.set_members(users)
+            except GroupException as e:
+                self.log_error(f"failed to set members of group '{group.name}' to only users {users}: {e}")
+                failed_repairs += 1
+            else:
+                self.log_info(f"members of group '{group.name}' set to only users {users}")
+                repairs += 1
+        else:
+            self.log_debug(f"group '{group.name}' does only contain members {users}")
+        
+        return repairs, failed_repairs
+
+
+class GroupException(Exception):
+    def __init__(self, message):
+        self.message = message
+
+
+class Group:
+    def __init__(self, name, gid, members):
+        self.name = name
+        self.gid = gid
+        self.members = members
+
+    
+    @staticmethod
+    def lookup(group):
+        try:
+            with open("/etc/group") as f:
+                for line in f:
+                    if line.startswith(group + ':'):
+                        entry = line.strip().split(':')
+                        name = entry[0]
+                        gid = entry[2]
+                        members = entry[3].split(',') if entry[3] else []
+                        return Group(name, gid, members)
+        except Exception as e:
+            raise GroupException(e)
+        return None
+        
+
+    @staticmethod
+    def create(name, gid=None):
+        command = ["groupadd", name]
+        if (gid):
+            command += ["--gid", gid]
+        process = Popen(command, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = process.communicate()
+
+        if (process.returncode != 0):
+            msg = stderr.decode("utf-8").strip().split('\n')[0]
```

Comment:
Why only the first line from the error output?

Reply:
My thoughts were to try to follow the logging guide lines in [CONTRIBUTING.md](https://github.com/cfengine/core/blob/master/CONTRIBUTING.md#logging-guidelines), stating "Normally, try to keep each message to one line of output, produced by one call to Log()."

Also the first line of the command line tools used here, clearly explains what's wrong. The remaining lines produced usually dwell into details on how to use that specific tool. This can be quite confusing to the user.

Reply:
Makes sense, but then it should be explained in a comment in the code.

Reply:
Good point :+1: